builder/googlecompute/block_device.go

```diff
@@ -0,0 +1,306 @@
+		errs = append(errs, fmt.Errorf("Valid values are SCSI or NVME"))
+	}
+
+	switch bd.AttachmentMode {
+	case "READ_ONLY", "READ_WRITE":
+	case "":
+		bd.AttachmentMode = "READ_WRITE"
+	default:
+		errs = append(errs, fmt.Errorf("Invalid attachment_mode: %q", bd.AttachmentMode))
+		errs = append(errs, fmt.Errorf("Valid values are READ_ONLY or READ_WRITE"))
+	}
+
+	if bd.DeviceName != "" && bd.VolumeType == LocalScratch {
+		errs = append(errs, fmt.Errorf("Scratch volumes may not have a device_name attached to them"))
+	}
+
+	if bd.SourceVolume != "" {
+		bd.KeepDevice = true
+	}
+
+	return errs
+}
+
+func (bd BlockDevice) hasDiskCreationArgs() bool {
+	return bd.VolumeSize != 0 ||
+		bd.VolumeType != "" ||
+		bd.DiskName != "" ||
+		bd.IOPS != 0 ||
+		bd.KeepDevice
+}
+
+func (bd *BlockDevice) prepareDiskCreate() []error {
+	if bd.SourceVolume != "" && bd.hasDiskCreationArgs() {
+		return []error{
+			fmt.Errorf(`if source_volume is set, none of the following arguments are settable:
+* disk_name
+* volume_type
+* volume_size
+* iops
+* keep_device`),
+		}
+	}
+
+	if bd.SourceVolume != "" {
+		return nil
+	}
+
+	var errs []error
+
+	switch bd.VolumeType {
+	case LocalScratch,
+		ZonalStandard, ZonalBalanced, ZonalSSD, ZonalExtreme:
+	case "":
+		errs = append(errs, fmt.Errorf("volume_type must be specified"))
+		errs = append(errs, fmt.Errorf("%s", volumeTypeError()))
+	default:
+		errs = append(errs, fmt.Errorf("Unknown volume type %q", bd.VolumeType))
+		errs = append(errs, fmt.Errorf("%s", volumeTypeError()))
+	}
+
+	if bd.VolumeType == LocalScratch && bd.KeepDevice {
+		errs = append(errs, fmt.Errorf("Scratch volumes cannot be kept after the instance is shutdown"))
+	}
+
+	if bd.VolumeType == LocalScratch && bd.DiskName != "" {
+		errs = append(errs, fmt.Errorf("Scratch volumes cannot have a name specified."))
+	}
+
+	if bd.VolumeSize == 0 {
+		errs = append(errs, fmt.Errorf("volume_size must be specified"))
+	}
+
+	if bd.DiskName != "" && (!diskNameRegex.MatchString(bd.DiskName) || len(bd.DiskName) > 63) {
+		errs = append(errs, fmt.Errorf("Disk name %q is non-compliant.", bd.DiskName))
+		errs = append(errs, fmt.Errorf("The name must be 1-63 characters long and comply to the regexp '[a-z]([-a-z0-9]*[a-z0-9])?'"))
+	}
+
+	if bd.VolumeType != LocalScratch && bd.DiskName == "" {
+		uuid, err := uuid.NewV4()
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error creating the disk name: %s", err))
+		} else {
+			bd.DiskName = fmt.Sprintf("packer-%s", uuid.String())
+			log.Printf("[TRACE] - Set disk name as %q", bd.DiskName)
+		}
+	}
+
+	if bd.IOPS != 0 && (bd.IOPS < 10000 || bd.IOPS > 120000) {
+		errs = append(errs, fmt.Errorf("Requested IOPS must be >= 10000 and <= 120000"))
+	}
+
+	if bd.IOPS != 0 {
+		switch bd.VolumeType {
+		case ZonalExtreme:
+		default:
+			errs = append(errs, fmt.Errorf("IOPS may only be specified for %q volumes", ZonalExtreme))
+		}
+	}
```

Comment:
Slight refactor to get rid of some additional conditions and group checks together. 
```go
	switch bd.VolumeType {
	case LocalScratch,
		ZonalStandard, ZonalBalanced, ZonalSSD, ZonalExtreme:
	default:
		errs = append(errs, fmt.Errorf("A valid volume type was not specified %q", bd.VolumeType))
		errs = append(errs, fmt.Errorf("%s", volumeTypeError()))
		// does it make sense to continue if we get this error? 
		return errs
	}
	
	if (bd.IOPS != 0) && (bd.VolumeType  != ZonalExtreme) {
		errs = append(errs, fmt.Errorf("IOPS may only be specified for %q volumes", ZonalExtreme))
	}
	
	if bd.IOPS != 0 && (bd.IOPS < 10000 || bd.IOPS > 120000) {
		errs = append(errs, fmt.Errorf("Requested IOPS must be >= 10000 and <= 120000"))
	}

	if bd.VolumeType == LocalScratch && bd.KeepDevice {
		errs = append(errs, fmt.Errorf("Scratch volumes cannot be kept after the instance is shutdown"))
	}

	if bd.VolumeType == LocalScratch && bd.DiskName != "" {
		errs = append(errs, fmt.Errorf("Scratch volumes cannot have a name specified."))
	}

	if bd.VolumeSize == 0 {
		errs = append(errs, fmt.Errorf("volume_size must be specified"))
	}
	
	//No need to continue checking for LocalScratch types
	if bd.VolumeType == LocalScratch {
	   return errs
	 }

	if bd.DiskName != "" && (!diskNameRegex.MatchString(bd.DiskName) || len(bd.DiskName) > 63) {
		errs = append(errs, fmt.Errorf("Disk name %q is non-compliant.", bd.DiskName))
		errs = append(errs, fmt.Errorf("The name must be 1-63 characters long and comply to the regexp '[a-z]([-a-z0-9]*[a-z0-9])?'"))
	}

       //If we've gotten here then we know VolumeType is not LocalScratch
	if bd.DiskName == "" {
		uuid, err := uuid.NewV4()
		if err != nil {
			errs = append(errs, fmt.Errorf("error creating the disk name: %s", err))
			return errs
		}
		bd.DiskName = fmt.Sprintf("packer-%s", uuid.String())
		log.Printf("[TRACE] - Set disk name as %q", bd.DiskName)
	}

```